src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,735 @@
+
+    For remote stores where the performance benefit is more apparent::
+
+        >>> from zarr.storage import RemoteStore
+        >>> # remote_store = RemoteStore.from_url("https://example.com/data.zarr")
+        >>> # cached_remote = zarr.LRUStoreCache(remote_store, max_size=2**28)
+
+
+    """
+
+    supports_writes: bool = True
+    supports_deletes: bool = True
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
+
+    root: Path
+
+    def __init__(self, store: Store, *, max_size: int, **kwargs: Any) -> None:
+        if max_size <= 0:
+            raise ValueError("max_size must be a positive integer (bytes)")
+
+        # Extract and handle known parameters
+        read_only = kwargs.get("read_only", getattr(store, "read_only", False))
+
+        # Call parent constructor with read_only parameter
+        super().__init__(read_only=read_only)
+
+        self._store = store
+        self._max_size = max_size
+        self._current_size = 0
+        self._keys_cache: list[str] | None = None
+        self._contains_cache: dict[Any, Any] = {}
+        self._listdir_cache: dict[str | None, list[str]] = {}
+        self._values_cache: OrderedDict[str, Any] = OrderedDict()
+        self._mutex = Lock()
+        self.hits = self.misses = 0
+
+        # Handle root attribute if present in underlying store
+        if hasattr(store, "root"):
+            self.root = store.root
+        else:
+            self.root = Path("/")  # Default root path
+
+    @classmethod
+    async def open(cls, store: Store, *, max_size: int, **kwargs: Any) -> "LRUStoreCache":
+        """
+        Create and open the LRU cache store.
+
+        Parameters
+        ----------
+        store : Store
+            The underlying store to wrap with caching.
+        max_size : int
+            The maximum size that the cache may grow to, in number of bytes.
+        **kwargs : Any
+            Additional keyword arguments passed to the store constructor.
+
+        Returns
+        -------
+        LRUStoreCache
+            The opened cache store instance.
+        """
+        cache = cls(store, max_size=max_size, **kwargs)
+        await cache._open()
+        return cache
+
+    def with_read_only(self, read_only: bool = False) -> "LRUStoreCache":
+        """
+        Return a new LRUStoreCache with a new read_only setting.
+
+        Parameters
+        ----------
+        read_only
+            If True, the store will be created in read-only mode. Defaults to False.
+
+        Returns
+        -------
+        LRUStoreCache
+            A new LRUStoreCache with the specified read_only setting.
+        """
+        # Create a new underlying store with the new read_only setting
+        underlying_store = self._store.with_read_only(read_only)
+        return LRUStoreCache(underlying_store, max_size=self._max_size, read_only=read_only)
+
+    def _normalize_key(self, key: Any) -> str:
```

Comment:
```suggestion
    def _normalize_key(self, key: str | Path) -> str:
```
Seems like the correct more tight typing for this?

Reply:
Also, this doesn't need to be a method because it doesn't use `self`, it can just live as a helper funciton somewhere.